core/src/main/scala/org/apache/spark/storage/BlockManagerMessages.scala

```diff
@@ -93,11 +94,15 @@ private[spark] object BlockManagerMessages {
 
   case class GetLocations(blockId: BlockId) extends ToBlockManagerMaster
 
-  case class GetLocationsAndStatus(blockId: BlockId) extends ToBlockManagerMaster
+  case class GetLocationsAndStatus(blockId: BlockId, requesterHost: String)
+    extends ToBlockManagerMaster
 
   // The response message of `GetLocationsAndStatus` request.
-  case class BlockLocationsAndStatus(locations: Seq[BlockManagerId], status: BlockStatus) {
-    assert(locations.nonEmpty)
+  case class BlockLocationsAndStatus(
+      locations: Seq[BlockManagerId],
+      status: BlockStatus,
+      localDirs: Option[Array[String]]) {
```

Comment:
the meaning of `localDirs` is a bit confusing, so its worth a comment.  IIUC, if it is persisted-to-disk on the same host as the original request, then it will be Some, and the cached data will be in a file in one of those dirs (which perhaps corresponds to dirs of a different executor); else, it will be None.